include/rmm/detail/logger.hpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <spdlog/sinks/basic_file_sink.h>
+#include <spdlog/spdlog.h>
+
+#include <iostream>
+#include <string>
+
+namespace rmm {
+
+namespace {
+
+inline std::string default_log_filename()
+{
+  auto filename = std::getenv("RMM_DEBUG_LOG_FILE");
+  return (filename == nullptr) ? std::string{"rmm_log.txt"} : std::string{filename};
+}
+}  // namespace
+
+inline spdlog::logger& logger()
+{
+  bool truncate_file{true};
+  static spdlog::logger logger_(
+    "RMM",
+    std::make_shared<spdlog::sinks::basic_file_sink_mt>(default_log_filename(), truncate_file));
+  static bool initialized{false};
+  if (not initialized) {
```

Comment:
There is a race condition here. C++11 function local static initialization will guarantee that only one thread will construct `initialized`, but after that many threads can potentially enter the `if(not initialized)` branch and race in executing the contents of the branch. 

I'd suggest putting construction of the `spdlog::logger` object and the below initialization (set_pattern, etc.) into the constructor of a wrapper object. Then we can just have a function local static instance of that wrapper object, e.g.,

```
inline spdlog::logger& logger(){
   static logger_wrapper w{...};
   return w.logger();
}
```

Then you can just rely on the construct once logic of the language instead of having to keep track of an `initialized` variable and potentially have races.